README
NAME
    Parse::CPAN::Meta - Parse META.yml and META.json CPAN metadata files

VERSION
    version 1.4414

SYNOPSIS
        #############################################
        # In your file
    
        ---
        name: My-Distribution
        version: 1.23
        resources:
          homepage: "http://example.com/dist/My-Distribution"
    
    
        #############################################
        # In your program
    
        use Parse::CPAN::Meta;
    
        my $distmeta = Parse::CPAN::Meta->load_file('META.yml');
    
        # Reading properties
        my $name     = $distmeta->{name};
        my $version  = $distmeta->{version};
        my $homepage = $distmeta->{resources}{homepage};

DESCRIPTION
    Parse::CPAN::Meta is a parser for META.json and META.yml files, using
    JSON::PP and/or CPAN::Meta::YAML.

    Parse::CPAN::Meta provides three methods: "load_file",
    "load_json_string", and "load_yaml_string". These will read and
    deserialize CPAN metafiles, and are described below in detail.

    Parse::CPAN::Meta provides a legacy API of only two functions, based on
    the YAML functions of the same name. Wherever possible, identical
    calling semantics are used. These may only be used with YAML sources.

    All error reporting is done with exceptions (die'ing).

    Note that META files are expected to be in UTF-8 encoding, only. When
    converted string data, it must first be decoded from UTF-8.

METHODS
  load_file
      my $metadata_structure = Parse::CPAN::Meta->load_file('META.json');

      my $metadata_structure = Parse::CPAN::Meta->load_file('META.yml');

    This method will read the named file and deserialize it to a data
    structure, determining whether it should be JSON or YAML based on the
    filename. The file will be read using the ":utf8" IO layer.

  load_yaml_string
      my $metadata_structure = Parse::CPAN::Meta->load_yaml_string($yaml_string);

    This method deserializes the given string of YAML and returns the first
    document in it. (CPAN metadata files should always have only one
    document.) If the source was UTF-8 encoded, the string must be decoded
    before calling "load_yaml_string".

  load_json_string
      my $metadata_structure = Parse::CPAN::Meta->load_json_string($json_string);

    This method deserializes the given string of JSON and the result. If the
    source was UTF-8 encoded, the string must be decoded before calling
    "load_json_string".

  load_string
      my $metadata_structure = Parse::CPAN::Meta->load_string($some_string);

    If you don't know whether a string contains YAML or JSON data, this
    method will use some heuristics and guess. If it can't tell, it assumes
    YAML.

  yaml_backend
      my $backend = Parse::CPAN::Meta->yaml_backend;

    Returns the module name of the YAML serializer. See "ENVIRONMENT" for
    details.

  json_backend
      my $backend = Parse::CPAN::Meta->json_backend;

    Returns the module name of the JSON serializer. This will either be
    JSON::PP or JSON. Even if "PERL_JSON_BACKEND" is set, this will return
    JSON as further delegation is handled by the JSON module. See
    "ENVIRONMENT" for details.

FUNCTIONS
    For maintenance clarity, no functions are exported by default. These
    functions are available for backwards compatibility only and are best
    avoided in favor of "load_file".

  Load
      my @yaml = Parse::CPAN::Meta::Load( $string );

    Parses a string containing a valid YAML stream into a list of Perl data
    structures.

  LoadFile
      my @yaml = Parse::CPAN::Meta::LoadFile( 'META.yml' );

    Reads the YAML stream from a file instead of a string.

ENVIRONMENT
  PERL_JSON_BACKEND
    By default, JSON::PP will be used for deserializing JSON data. If the
    "PERL_JSON_BACKEND" environment variable exists, is true and is not
    "JSON::PP", then the JSON module (version 2.5 or greater) will be loaded
    and used to interpret "PERL_JSON_BACKEND". If JSON is not installed or
    is too old, an exception will be thrown.

  PERL_YAML_BACKEND
    By default, CPAN::Meta::YAML will be used for deserializing YAML data.
    If the "PERL_YAML_BACKEND" environment variable is defined, then it is
    interpreted as a module to use for deserialization. The given module
    must be installed, must load correctly and must implement the "Load()"
    function or an exception will be thrown.

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Parse-CPAN-Meta>. You
    will be notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/Perl-Toolchain-Gang/Parse-CPAN-Meta>

      git clone https://github.com/Perl-Toolchain-Gang/Parse-CPAN-Meta.git

AUTHORS
    *   Adam Kennedy <adamk@cpan.org>

    *   David Golden <dagolden@cpan.org>

CONTRIBUTORS
    *   Graham Knop <haarg@haarg.org>

    *   Joshua ben Jore <jjore@cpan.org>

    *   Neil Bowers <neil@bowers.com>

    *   Ricardo Signes <rjbs@cpan.org>

    *   Steffen Mueller <smueller@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Adam Kennedy and Contributors.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changes for Perl programming language extension Parse-CPAN-Meta

1.4414    2014-03-04 15:22:00-05:00 America/New_York

    [ADDED]

    - Added load_string, which uses heuristics to guess YAML/JSON content

    [CHANGED]

    - load_file will now attempt to detect YAML/JSON content if the suffix
      is not .json, .yml or .yaml

1.4413    2014-02-17 20:04:23-05:00 America/New_York

    [FIXED]

    - UTF-8 decoding is done differently to avoid requiring
      a newer version of Encode (Graham Knop)

1.4412    2014-02-17 07:14:43-05:00 America/New_York

    [FIXED]

    - Added prerequisite on Encode 2.11 (Paul Howarth)

1.4411    2014-02-16 23:06:00-05:00 America/New_York

    [FIXED]

    - Invalid UTF-8 encoding in YAML files are now replaced with "PERLQQ"
      quoting from the Encode module and without warnings.

    - Removed legacy test modifications for $ENV{PERL_CORE}

1.4410    2014-02-16 00:47:48-05:00 America/New_York

    [INCOMPATIBLE CHANGES]

    - Dropped support for Perl's before v5.8.1 to ensure proper
      UTF-8 and IO layer support

    [CHANGED]

    - Updated CPAN::Meta::YAML prereq to 0.011 due to deprecation
      of 'errstr' and consistent use of exceptions

1.4409    2013-09-23 13:17:18 America/New_York

    [FIXED]

    - updated Makefile.PL logic to support PERL_NO_HIGHLANDER

1.4408    2013-09-23 11:06:35 America/New_York

    [FIXED]

    - Installation on Perls < 5.12 will uninstall older versions installed
      due to being bundled with ExtUtils::MakeMaker

1.4407    2013-09-08 10:16:29 America/New_York

    - Fixed incorrect resource metadata

1.4406    2013-09-08 09:59:28 America/New_York

    - Fixed improperly encoded META.yml

1.4405    2013-07-17 21:43:34 America/New_York

    - Fixed incorrect "return ... or die ..." constructs

    - Converted distribution to Dist::Zilla management

    - Noted new repository location in the Github Perl-Toolchain-Gang
      organization

1.4404 2012-04-05

    - Protected tests from user PERL_YAML/JSON_BACKEND

1.4403 2012-04-01

    - Bumped prereqs: JSON::PP to 2.27200 and CPAN::Meta::YAML to 0.008

1.4402 2012-02-07

    - Minor maintenance: standardized newlines to Unix style

1.4401 2011-02-04

    - Removed Module::Load::Conditional dependency

1.4400 2011-02-04

    - Added 'json_backend' and 'yaml_backend' methods to provide
    the names of modules used for deserialization

1.4200 2011-01-24

    - No changes from 1.41_04

1.41_04 2011-01-03

    - Support PERL_JSON_BACKEND environment (defaulting to JSON::PP)

    - Support PERL_YAML_BACKEND environment (defaulting to CPAN::Meta:YAML)

    - Update Makefile.PL to install to sitelib on Perl 5.12+

1.41_03 2010-12-17

    - Throw exception when CPAN::Meta::YAML has a parse error
      (RT#47608)

1.41_02 2010-12-17

    - Convert to using CPAN::Meta::YAML instead of YAML::Tiny

1.41_01 2010-12-10

    - add support for JSON metafiles and load_* methods (RJBS)

    - Move binary file unpacking to test file from Makefile.PL so tests
      will pass in the Perl core using a generated Makefile.PL

    - Move bundled uupacktool.pl to t/bin/

1.40 2009-07-25

    - Add core perl 5.10.1's uupacktool.pl

    - Repackage t/data/utf_16_le_bom.yml as ASCII for https://rt.cpan.org/Ticket/Display.html?id=47844

1.39 2009-05-21

    - Even though utf8 starts at 5.7+ there's no is_utf till
      5.8.1 so skip in the tests if needed (ADAMK)

1.38 2009-05-16

    - Synchorised version numbers with YAML::Tiny

    - Code shrinkages to the bloaty Unicode/BOM code

    - Merging a ton more tests from YAML::Tiny

    - Added back some pointless options to yaml_ok that keep our
      codebase closer to YAML::Tiny (which should help with syncing)

    - Changed exceptions to be closer to YAML::Tiny

0.05 2009-03-11

    - Upgrade previous release to a stable release.

0.04_01 2009-03-03

    - Merge Unicode / BOM fixes from YAML::Tiny

0.04 2009-01-07

    - Matching changes in YAML::Tiny 1.36

    - Fixing missing feature reported by H.Merijn Brand

    - Changes to make Padre-CPAN-Meta core-compatible (SMUELLER)++

0.03 2008-03-20

    - Adding YAML.pm compatibility in scalar context

0.02 2008-01-10

    - Purging some references to YAML::Tiny that I missed

0.01 2008-01-08

    - Cloned from YAML::Tiny 1.21

    - Removed all write functionality

    - Reduced interface to Load and LoadFile

    - Removed object-orientation

    - Removed global $errstr, all errors are exceptions

    - Applied prototypes to all functions


lib/Parse/CPAN/Meta.pm
use 5.008001;
use strict;
package Parse::CPAN::Meta;
# ABSTRACT: Parse META.yml and META.json CPAN metadata files
our $VERSION = '1.4411'; # VERSION

use Exporter;
use Carp 'croak';

our @ISA = qw/Exporter/;
our @EXPORT_OK = qw/Load LoadFile/;

sub load_file {
  my ($class, $filename) = @_;

  if ($filename =~ /\.ya?ml$/) {
    return $class->load_yaml_string(_slurp($filename));
  }

  if ($filename =~ /\.json$/) {
    return $class->load_json_string(_slurp($filename));
  }

  croak("file type cannot be determined by filename");
}

sub load_yaml_string {
  my ($class, $string) = @_;
  my $backend = $class->yaml_backend();
  my $data = eval { no strict 'refs'; &{"$backend\::Load"}($string) };
  croak $@ if $@;
  return $data || {}; # in case document was valid but empty
}

sub load_json_string {
  my ($class, $string) = @_;
  my $data = eval { $class->json_backend()->new->decode($string) };
  croak $@ if $@;
  return $data || {};
}

sub yaml_backend {
  if (! defined $ENV{PERL_YAML_BACKEND} ) {
    _can_load( 'CPAN::Meta::YAML', 0.011 )
      or croak "CPAN::Meta::YAML 0.011 is not available\n";
    return "CPAN::Meta::YAML";
  }
  else {
    my $backend = $ENV{PERL_YAML_BACKEND};
    _can_load( $backend )
      or croak "Could not load PERL_YAML_BACKEND '$backend'\n";
    $backend->can("Load")
      or croak "PERL_YAML_BACKEND '$backend' does not implement Load()\n";
    return $backend;
  }
}

sub json_backend {
  if (! $ENV{PERL_JSON_BACKEND} or $ENV{PERL_JSON_BACKEND} eq 'JSON::PP') {
    _can_load( 'JSON::PP' => 2.27103 )
      or croak "JSON::PP 2.27103 is not available\n";
    return 'JSON::PP';
  }
  else {
    _can_load( 'JSON' => 2.5 )
      or croak  "JSON 2.5 is required for " .
                "\$ENV{PERL_JSON_BACKEND} = '$ENV{PERL_JSON_BACKEND}'\n";
    return "JSON";
  }
}

sub _slurp {
  require Encode;
  require PerlIO::encoding;
  local $PerlIO::encoding::fallback = Encode::PERLQQ()|Encode::STOP_AT_PARTIAL();
  open my $fh, "<:encoding(UTF-8)", "$_[0]" ## no critic
    or die "can't open $_[0] for reading: $!";
  return do { local $/; <$fh> };
}
  
sub _can_load {
  my ($module, $version) = @_;
  (my $file = $module) =~ s{::}{/}g;
  $file .= ".pm";
  return 1 if $INC{$file};
  return 0 if exists $INC{$file}; # prior load failed
  eval { require $file; 1 }
    or return 0;
  if ( defined $version ) {
    eval { $module->VERSION($version); 1 }
      or return 0;
  }
  return 1;
}

# Kept for backwards compatibility only
# Create an object from a file
sub LoadFile ($) {
  return Load(_slurp(shift));
}

# Parse a document from a string.
sub Load ($) {
  require CPAN::Meta::YAML;
  my $object = eval { CPAN::Meta::YAML::Load(shift) };
  croak $@ if $@;
  return $object;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Parse::CPAN::Meta - Parse META.yml and META.json CPAN metadata files

=head1 VERSION

version 1.4411

=head1 SYNOPSIS

    #############################################
    # In your file
    
    ---
    name: My-Distribution
    version: 1.23
    resources:
      homepage: "http://example.com/dist/My-Distribution"
    
    
    #############################################
    # In your program
    
    use Parse::CPAN::Meta;
    
    my $distmeta = Parse::CPAN::Meta->load_file('META.yml');
    
    # Reading properties
    my $name     = $distmeta->{name};
    my $version  = $distmeta->{version};
    my $homepage = $distmeta->{resources}{homepage};

=head1 DESCRIPTION

B<Parse::CPAN::Meta> is a parser for F<META.json> and F<META.yml> files, using
L<JSON::PP> and/or L<CPAN::Meta::YAML>.

B<Parse::CPAN::Meta> provides three methods: C<load_file>, C<load_json_string>,
and C<load_yaml_string>.  These will read and deserialize CPAN metafiles, and
are described below in detail.

B<Parse::CPAN::Meta> provides a legacy API of only two functions,
based on the YAML functions of the same name. Wherever possible,
identical calling semantics are used.  These may only be used with YAML sources.

All error reporting is done with exceptions (die'ing).

Note that META files are expected to be in UTF-8 encoding, only.  When
converted string data, it must first be decoded from UTF-8.

=begin Pod::Coverage




=end Pod::Coverage

=head1 METHODS

=head2 load_file

  my $metadata_structure = Parse::CPAN::Meta->load_file('META.json');

  my $metadata_structure = Parse::CPAN::Meta->load_file('META.yml');

This method will read the named file and deserialize it to a data structure,
determining whether it should be JSON or YAML based on the filename.
The file will be read using the ":utf8" IO layer.

=head2 load_yaml_string

  my $metadata_structure = Parse::CPAN::Meta->load_yaml_string($yaml_string);

This method deserializes the given string of YAML and returns the first
document in it.  (CPAN metadata files should always have only one document.)
If the source was UTF-8 encoded, the string must be decoded before calling
C<load_yaml_string>.

=head2 load_json_string

  my $metadata_structure = Parse::CPAN::Meta->load_json_string($json_string);

This method deserializes the given string of JSON and the result.  
If the source was UTF-8 encoded, the string must be decoded before calling
C<load_json_string>.

=head2 yaml_backend

  my $backend = Parse::CPAN::Meta->yaml_backend;

Returns the module name of the YAML serializer. See L</ENVIRONMENT>
for details.

=head2 json_backend

  my $backend = Parse::CPAN::Meta->json_backend;

Returns the module name of the JSON serializer.  This will either
be L<JSON::PP> or L<JSON>.  Even if C<PERL_JSON_BACKEND> is set,
this will return L<JSON> as further delegation is handled by
the L<JSON> module.  See L</ENVIRONMENT> for details.

=head1 FUNCTIONS

For maintenance clarity, no functions are exported by default.  These functions
are available for backwards compatibility only and are best avoided in favor of
C<load_file>.

=head2 Load

  my @yaml = Parse::CPAN::Meta::Load( $string );

Parses a string containing a valid YAML stream into a list of Perl data
structures.

=head2 LoadFile

  my @yaml = Parse::CPAN::Meta::LoadFile( 'META.yml' );

Reads the YAML stream from a file instead of a string.

=head1 ENVIRONMENT

=head2 PERL_JSON_BACKEND

By default, L<JSON::PP> will be used for deserializing JSON data. If the
C<PERL_JSON_BACKEND> environment variable exists, is true and is not
"JSON::PP", then the L<JSON> module (version 2.5 or greater) will be loaded and
used to interpret C<PERL_JSON_BACKEND>.  If L<JSON> is not installed or is too
old, an exception will be thrown.

=head2 PERL_YAML_BACKEND

By default, L<CPAN::Meta::YAML> will be used for deserializing YAML data. If
the C<PERL_YAML_BACKEND> environment variable is defined, then it is interpreted
as a module to use for deserialization.  The given module must be installed,
must load correctly and must implement the C<Load()> function or an exception
will be thrown.

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<http://rt.cpan.org/Public/Dist/Display.html?Name=Parse-CPAN-Meta>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/Perl-Toolchain-Gang/Parse-CPAN-Meta>

  git clone https://github.com/Perl-Toolchain-Gang/Parse-CPAN-Meta.git

=head1 AUTHORS

=over 4

=item *

Adam Kennedy <adamk@cpan.org>

=item *

David Golden <dagolden@cpan.org>

=back

=head1 CONTRIBUTORS

=over 4

=item *

Joshua ben Jore <jjore@cpan.org>

=item *

Neil Bowers <neil@bowers.com>

=item *

Ricardo Signes <rjbs@cpan.org>

=item *

Steffen Mueller <smueller@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Adam Kennedy and Contributors.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.013.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Parse META.yml and META.json CPAN metadata files",
  "AUTHOR" => "Adam Kennedy <adamk\@cpan.org>, David Golden <dagolden\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Parse-CPAN-Meta",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Parse::CPAN::Meta",
  "PREREQ_PM" => {
    "CPAN::Meta::YAML" => "0.011",
    "Carp" => 0,
    "Encode" => 0,
    "Exporter" => 0,
    "File::Spec" => "0.80",
    "JSON::PP" => "2.27200",
    "strict" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => "0.80",
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Test::More" => "0.47",
    "lib" => 0,
    "vars" => 0,
    "version" => 0,
    "warnings" => 0
  },
  "VERSION" => "1.4414",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::Meta::YAML" => "0.011",
  "Carp" => 0,
  "Encode" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => "0.80",
  "File::Spec::Functions" => 0,
  "JSON::PP" => "2.27200",
  "List::Util" => 0,
  "Test::More" => "0.47",
  "lib" => 0,
  "strict" => 0,
  "vars" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

# Added by Dist::Zilla::Plugin::MakeMaker::Highlander
if ( $] < 5.012
  && ! $ENV{PERL_NO_HIGHLANDER}
  && ! ( $ENV{PERL_MM_OPT} && $ENV{PERL_MM_OPT} =~ /(?:INSTALL_BASE|PREFIX)/ )
  && ! grep { /INSTALL_BASE/ || /PREFIX/ } @ARGV
) {
    $WriteMakefileArgs{UNINST} = 1;
}

$WriteMakefileArgs{INSTALLDIRS} = 'perl'
    if $] >= 5.010001 && $] <= 5.011000;

WriteMakefile(%WriteMakefileArgs);



